### PR TITLE
Fix #14967 - 

### LINKED ISSUE #14967: WC Logger Entries Not Ordering Correctly
### EXPLANATION OF THE ISSUE

In the WC Logger, log entries shown from the database interface are not displaying in true chronological order as the events are fired. This occurs because the events are ordering by the second in the timestamp. If the timestamps were order by milliseconds the events should show in the actual sequence as they occur. 

### STEPS TO REPRODUCE THE ISSUE
Create several wc_logger events that run sequentially in the codebase in close timestamp proximity. The WC Log in the wp admin will show those events out of order if they occur within the same second.

### PR DESCRIPTION
**Pull request by reesscot**
Allows logs to be stored by microseconds by moving the responsibility of the timestamp to the database.

Fixes #14967

### CODE REVIEW

**Comment by claudiosanches** [comment]
This will require a update script and maybe could be some difference for database timezone.

**Comment by sirreal** [comment]
> This will require a update script

👍 

> maybe could be some difference for database timezone.

@claudiosanches My understanding is that this should just add precision to the column, why would timezone be affected?

File: includes/class-wc-install.php [hunk 1]
```diff
@@ -605,7 +605,7 @@ private static function get_schema() {
 ) $collate;
 CREATE TABLE {$wpdb->prefix}woocommerce_log (
   log_id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
-  timestamp datetime NOT NULL,
+  timestamp datetime(6) NOT NULL DEFAULT NOW(5),
```
Comment (sirreal) [inline includes/class-wc-install.php]: This is going to be problematic as it appears to be _invalid syntax_ in MySQL `<5.6`.

You can see it failing in the automated tests:

> WordPress database error You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(6) NOT NULL DEFAULT NOW(5),

https://travis-ci.org/woocommerce/woocommerce/jobs/230864195#L316

Maybe a [version comment](https://dev.mysql.com/doc/refman/5.5/en/comments.html) will work here, something like `timestamp datetime/*!50600 (6) */ NOT NULL DEFAULT NOW/*!50600 (5) */` (untested).

File: includes/log-handlers/class-wc-log-handler-db.php [hunk 2]
```diff
@@ -55,18 +54,16 @@ public function handle( $timestamp, $level, $message, $context ) {
 	 *
 	 * @return bool True if write was successful.
 	 */
-	protected static function add( $timestamp, $level, $message, $source, $context ) {
+	protected static function add($level, $message, $source, $context ) {
```
Comment (sirreal) [inline includes/log-handlers/class-wc-log-handler-db.php]: Style nit: space inside the parens 🙂

**Review by sirreal — CHANGES_REQUESTED** [review]
Thanks for the PR! Unfortunately, this only fixes for MySQL 5.6+ and actually appears to _break_ MySQL <5.6 as-is!

I think we can get this working in a more compatible way, please see my comments for a suggestion. I'm happy to look again when this issue is ironed out.

The downside of this approach is that it will only fix for MySQL 5.6+. Any system using an inferior version will continue to have the issue.

**Comment by claudiosanches** [comment]
@sirreal 

> @claudiosanches My understanding is that this should just add precision to the column, why would timezone be affected?

Currently the `$timestamp` comes from:

```php
	public function log( $level, $message, $context = array() ) {
		if ( ! WC_Log_Levels::is_valid_level( $level ) ) {
			wc_doing_it_wrong( __METHOD__, sprintf( __( 'WC_Logger::log was called with an invalid level "%s".', 'woocommerce' ), $level ), '3.0' );
		}

		if ( $this->should_handle( $level ) ) {
			$timestamp = current_time( 'timestamp' );

			foreach ( $this->handlers as $handler ) {
				$handler->handle( $timestamp, $level, $message, $context );
			}
		}
	}
```

As you can see in https://developer.wordpress.org/reference/functions/current_time/ the timestamp uses the WordPress options to determine the timezone, that is not always the same timezone from the server or database.

**Comment by sirreal** [comment]
@claudiosanches You're right! Nice catch 👍

**Comment by claudiosanches** [comment]
Not to mention that this change has not been tested, since the `WC_Log_Handler_Interface` expects a `$timezone` has first parameter in the `handle` method.
If change from database, need to change for the email and file handlers too.
Better find other solution for it.

**Pull request closed without merging**